axolotls/numeric_column.py

```diff
@@ -133,14 +133,31 @@ def presence(self) -> Optional[torch.BoolTensor]:
         return self._presence
 
     def __len__(self) -> int:
-        return len(self._values)
+        return self._values.numel()
 
     def to_arrow(self):
         # TODO: Check whether PyArrow is available 
         import pyarrow as pa
 
-        values = self.values.numpy()
-        mask = ~self.presence.numpy() if self.presence is not None else None
+        if not self.presence is not None and self.values.dtype != torch.bool:
+            # Wrap Tensor memory into Arrow buffer, this avoids dependency on NumPy
+            values_buffer = pa.foreign_buffer(
```

Comment:
@dracifer , @ejguan :  . Directly construct Arrow Buffer from PyTorch Tensor's memory ^_^

Reply:
PyTorch doesn't yet support Python buffer protocol, see, for example:
* https://github.com/pytorch/pytorch/issues/69491#issuecomment-988482619
* https://pearu.github.io/array_interface_pytorch.html

Otherwise, we may be able to directly construct PyArrow buffer from the object that implements Python buffer protocol.

Reply:
Wow. Looks great.

Reply:
Do you need to validate the values being contiguous?

Reply:
@ejguan  Good point. Yeah we should validate it's continuous :)